source/dhtslib/sam.d

```diff
@@ -178,43 +216,167 @@ class SAMRecord
         {
             s[i] = seq_nt16_str[bam_seqi(seqdata, i)];
         }
-        return s;
+        return s.idup;
+    }
+
+    /// Assigns sequence and resets quality score
+    pragma(inline, true)
+    @property void sequence(string seq)
+    {
+        //nibble encode sequence
+        ubyte[] en_seq;
+        en_seq.length=(seq.length+1)>>1;
+        for(auto i=0;i<seq.length;i++){
+            en_seq[i>>1]|=seq_nt16_table[seq[i]]<< ((~i&1)<<2);
+        }
+
+        //get length of data before seq
+        uint l_prev=b.core.l_qname + cast(uint)(b.core.n_cigar*uint.sizeof);
+
+        //get starting point after seq
+        uint start_rest=l_prev+((b.core.l_qseq+1)>>1)+b.core.l_qseq;
+
+        //copy previous data
+        ubyte[] prev=b.data[0..l_prev].dup;
+
+        //copy rest of data
+        ubyte[] rest=b.data[start_rest..b.l_data].dup;
+        
+        //if not enough space
+        if(en_seq.length+seq.length+l_prev+(b.l_data-start_rest)>b.m_data){
+            auto ptr=cast(ubyte*)realloc(b.data,en_seq.length+seq.length+l_prev+(b.l_data-start_rest));
+            assert(ptr!=null);
+            b.data=ptr;
+            b.m_data=cast(uint)en_seq.length+cast(uint)seq.length+l_prev+(b.l_data-start_rest);
+        }
+
+        //reassign q_name and rest of data
+        b.data[0..l_prev]=prev;
+        b.data[l_prev..en_seq.length+l_prev]=en_seq;
+        //set qscores to 255 
+        memset(b.data+en_seq.length+l_prev,255,seq.length);
+        b.data[l_prev+en_seq.length+seq.length..l_prev+en_seq.length+seq.length+(b.l_data-start_rest)]=rest;
+
+        //reset data length, seq length
+        b.l_data=cast(uint)en_seq.length+cast(uint)seq.length+l_prev+(b.l_data-start_rest);
+        b.core.l_qseq=cast(int)(seq.length);
     }
 
     /// Return pointer to quality Cstring
     /// see samtools/sam_view.c: get_quality
-    @property char* qscores()
+    char[] qscores(bool phredScale=true)()
     {
         // calloc fills with \0; +1 len for Cstring
-        char* q = cast(char*) calloc(1, this.b.core.l_qseq + 1);
-        //char[] q;
-        //q.length = this.b.core.l_qseq;
+        // char* q = cast(char*) calloc(1, this.b.core.l_qseq + 1);
+        char[] q;
+        q.length = this.b.core.l_qseq;
 
         // auto bam_get_qual(bam1_t *b) { return (*b).data + ((*b).core.n_cigar<<2) + (*b).core.l_qname + (((*b).core.l_qseq + 1)>>1); }
         char* qualdata = cast(char*) bam_get_qual(this.b);
 
         for (int i; i < this.b.core.l_qseq; i++)
-            q[i] = cast(char)(qualdata[i] + 33);
+            static if(phredScale) q[i] = cast(char)(qualdata[i] + 33);
+            else q[i] = cast(char)(qualdata[i]);
         return q;
     }
 
+    ///Adds qscore sequence given that it is the same length as the bam sequence
+    pragma(inline, true)
+    void q_scores(bool phredScale=true)(string seq)
+    {
+        if(seq.length!=b.core.l_qseq){
```

Comment:
Question for discussion (not blocking): What if we wrapped checks like this in conditional compilation such that you could have a version(superfastunsafe) (or whatever) that skipped this and other checks, but by default they are compiled in?

Reply:
Could definitely be useful. If you want raw speed though might be best to just use C API.